### PR TITLE
feat(api,ui): websocket server + client for admin

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/websocket.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// Subscribe to project
+	if (m.ProjectKey != "" && m.WorkflowName == "") || (m.ProjectKey != "" && m.Operation != "") {
+		perms, err := permission.LoadProjectMaxLevelPermission(ctx, db, []string{m.ProjectKey}, getAPIConsumer(ctx).GetGroupIDs())
+		if err != nil {
+			return err
+		}
+		maxLevelPermission := perms.Level(m.ProjectKey)
+		if maxLevelPermission < sdk.PermissionRead && !isMaintainer(ctx) && !isAdmin(ctx) {
```
Comment (richardlt) [inline engine/api/websocket.go:233]: For now there is nothing that prevent to create a consumer for a hatchery that is maintainer or admin. But is this case we don't want to send all event but only ones based on groups ids.

File: engine/api/websocket.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	// Subscribe to workflow
+	if m.ProjectKey != "" && m.WorkflowName != "" {
+		perms, err := permission.LoadWorkflowMaxLevelPermission(ctx, db, m.ProjectKey, []string{m.WorkflowName}, getAPIConsumer(ctx).GetGroupIDs())
+		if err != nil {
+			return err
+		}
+		maxLevelPermission := perms.Level(m.WorkflowName)
+		if maxLevelPermission < sdk.PermissionRead && !isMaintainer(ctx) && !isAdmin(ctx) {
```
Comment (richardlt) [inline engine/api/websocket.go:246]: Same

File: engine/api/event/publish_workflow_run.go [hunk 3]
```diff
@@ -6,15 +6,27 @@ import (
...
 	"github.com/ovh/cds/sdk/log"
 )
 
-func publishRunWorkflow(ctx context.Context, payload interface{}, key, workflowName, appName, pipName, envName string, num int64, sub int64, status string, tags []sdk.WorkflowRunTag, eventIntegrations []sdk.ProjectIntegration) {
-	eventIntegrationsID := make([]int64, len(eventIntegrations))
-	for i, eventIntegration := range eventIntegrations {
+type publishWorkflowRunData struct {
+	projectKey        string
+	workflowName      string
+	applicationName   string
```
Comment (fsamin) [inline engine/api/event/publish_workflow_run.go:16]: i think applicationName, environmentName, pipelineName and EventIntegrationsID  doesn't have sens in this event

File: engine/api/event/publish_workflow_run.go [hunk 4]
```diff
@@ -157,19 +178,45 @@ func PublishWorkflowNodeRun(ctx context.Context, nr sdk.WorkflowNodeRun, w sdk.W
...
 }
 
 // PublishWorkflowNodeJobRun publish a WorkflowNodeJobRun
-func PublishWorkflowNodeJobRun(ctx context.Context, db gorp.SqlExecutor, pkey string, wr sdk.WorkflowRun, jr sdk.WorkflowNodeJobRun) {
+func PublishWorkflowNodeJobRun(ctx context.Context, pkey string, wr sdk.WorkflowRun, jr sdk.WorkflowNodeJobRun) {
 	e := sdk.EventRunWorkflowJob{
-		ID:     jr.ID,
-		Status: jr.Status,
-		Start:  jr.Start.Unix(),
+		ID:           jr.ID,
```
Comment (fsamin) [inline engine/api/event/publish_workflow_run.go:199]: Is it possible to add hatcheryName and Workermodel in this event ?

File: engine/api/websocket.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+		if r := recover(); r != nil {
+			err = fmt.Errorf("websocketClient.Send recovered %v", r)
+		}
+	}()
+
+	if c == nil || c.con == nil || !c.isAlive.IsSet() {
+		return nil
+	}
+
+	if event.EventType == fmt.Sprintf("%T", sdk.EventRunWorkflowJob{}) && c.filter.Queue {
```
Comment (fsamin) [inline engine/api/websocket.go:271]: a switch statement would be nice

File: engine/api/api_routes.go [hunk 6]
```diff
@@ -405,6 +416,7 @@ func (api *API) InitRouter() {
 
 	// SSE
 	r.Handle("/events", ScopeNone(), r.GET(api.eventsBroker.ServeHTTP))
+	r.Handle("/ws", ScopeNone(), r.GET(api.websocketBroker.ServeHTTP))
```
Comment (fsamin) [inline engine/api/api_routes.go:419]: It is probably not so easy but we absolutely need unit test, i remember we had so much trouble with eventsBroker panics every where.

We have to check, connection, disconnection from the web socket and checks that all the references are cleaned on the broker side.
Also all the suscription and the filter feature.

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13041.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13045.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Pull request merged**